### PR TITLE
[SYCL] Refactor USM allocator to improve ABI stability

### PR DESCRIPTION
**Pull request by alexbatashev**
This patch is a part of effort to decouple SYCL Runtime library
interface from its actual implementation. The goal is to improve
SYCL ABI/API compatibility between different versions of library.

USM allocator was refactored to use public C-style API instead of
private C++-style APIs. Also, documentation was improved.

Signed-off-by: Alexander Batashev <alexander.batashev@intel.com>

### CODE REVIEW

File: sycl/include/CL/sycl/usm/usm_allocator.hpp [hunk 1]
```diff
@@ -107,35 +122,26 @@ class usm_allocator {
         "Device pointers do not support address on host");
   }
 
-  // Allocate memory
-  template <
-      usm::alloc AllocT = AllocKind,
-      typename std::enable_if<AllocT == usm::alloc::host, int>::type = 0>
+  /// Allocates memory.
+  ///
+  /// @param Size is a count of elements to allocate memory for.
```
Comment (bader) [inline sycl/include/CL/sycl/usm/usm_allocator.hpp]: Wondering if `Size` is the right name for the parameter, which is `a count of elements ...`.
`NumOfElements`?

File: sycl/include/CL/sycl/usm/usm_allocator.hpp [hunk 2]
```diff
@@ -107,35 +122,26 @@ class usm_allocator {
...
     }
     return Result;
   }
 
-  // Deallocate memory
-  void deallocate(pointer Ptr, size_t size) {
+  /// Deallocates memory.
+  ///
+  /// @param Ptr is a pointer to memory being deallocated.
+  void deallocate(pointer Ptr, size_t Size) {
```
Comment (bader) [inline sycl/include/CL/sycl/usm/usm_allocator.hpp]: `Size` is not documented.
What do we need `Size` for?
Is it possible to deallocate a part of the previously allocated memory?

Comment (jbrodman) [reply sycl/include/CL/sycl/usm/usm_allocator.hpp]: https://en.cppreference.com/w/cpp/memory/allocator/deallocate

Symmetry?

Comment (alexbatashev) [reply sycl/include/CL/sycl/usm/usm_allocator.hpp]: @bader cppreference says:
> **n** - number of objects earlier passed to allocate

I'm not sure if this method is designed to do partial deallocations. @jbrodman can you please clarify?

**Pull request merged**